crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1278,14 +1278,17 @@ impl<'db> Type<'db> {
 
             (Type::SubclassOf(_), Type::SubclassOf(_)) => false,
 
-            (Type::SubclassOf(_), Type::Instance(instance))
-            | (Type::Instance(instance), Type::SubclassOf(_)) => {
-                // TODO this should be `true` if the instance is of a final type which is not a
-                // subclass of type. (With non-final types, we never know whether a subclass might
-                // multiply-inherit `type` or a subclass of it, and thus not be disjoint with
-                // `type[...]`.) Until we support finality, hardcode None, which is known to be
-                // final.
-                instance.class.is_known(db, KnownClass::NoneType)
+            (Type::SubclassOf(subclass_of_ty), instance @ Type::Instance(_))
+            | (instance @ Type::Instance(_), Type::SubclassOf(subclass_of_ty)) => {
+                // `type[T]` is disjoint from `S`, where `S` is an instance type,
+                // if `U` is disjoint from `S`,
+                // where `U` represents all instances of `T`'s metaclass
+                let metaclass_instance = subclass_of_ty
+                    .subclass_of()
+                    .into_class()
+                    .map(|class| class.metaclass(db).to_instance(db))
+                    .unwrap_or_else(|| KnownClass::Type.to_instance(db));
+                instance.is_disjoint_from(db, metaclass_instance)
```

Comment:
Nice! I like that we can delegate to instance-vs-instance, rather than checking finality here also.

It seems right that `T` actually disappears from this check, apart from its metaclass. There's no difference between `type[object]` and `type[OtherClass]` as regards disjointness from an instance type, as long as the metaclass of `OtherClass` is just `type`.